test/unit/validation.test.ts

```diff
@@ -120,24 +114,20 @@ describe('Validation', () => {
     }
 
     beforeEach(() => {
-        // Define the class for each test, so that the schema is re-created every time.
-        class LoginClass {
-            @string()
-            emailAddress?: string;
-
-            @string()
-            password?: string;
-        }
-        login = new LoginClass();
+        Login = getLoginClass();
+        login = new Login();
         login.emailAddress = 'joe@example.com';
-        Login = LoginClass;
-        mockJoi = createMockJoi();
-    });
-
-    afterEach(() => {
-        mockJoi.array.mockReset();
-        mockJoi.object.mockReset();
-        mockJoi.validate.mockReset();
+        joi = partialOf<typeof Joi>({
```

Comment:
I use `partialOf` in most of my TS tests. It's a just ridiculously simple but handy function for enforcing that your mock is mocking actual properties of your target type and that they are implemented correctly, and then pretends to return the entire target type. Useful just for testing purposes.

Reply:
That looks handy. Might need a bit more type finagling to support the mock function properties. 

In the changes I made, I wanted to see if we could make it type safe all the way through from tests to main implementation. This included some changes so that functions required only a subset of `Joi`'s methods, instead of a complete `Joi` insdtance. I think now, using `partialOf()`, if you were to remove the mocked `array` property, it wouldn't error at compile time, only runtime. (Not a big problem since it's test code.)

If we don't want that level of type-safety between the tests and the main code, should we revert the changes I made to `validation.ts` and `getJoiSchema()`? I changed their signatures so that the `joi` arguments only needed the minimum required properties (`array`, `object`, `validate`).

Reply:
> If we don't want that level of type-safety between the tests and the main code, should we revert the changes I made to validation.ts and getJoiSchema()?

Up to you. :) I hear you regarding complete type-safety in tests. I don't have a strong opinion on it as I do this with my tests and have gotten used to it. As you say, if somethings not implemented you usually know about it pretty quick because your tests fail. But yeah if you prefer you can undo my commit and just take any bits you like.

Reply:
Alright, I think let's keep the current changes as-is.